processor/metricstarttimeprocessor/internal/datapointstorage/cache.go

```diff
@@ -0,0 +1,115 @@
+// Copyright The OpenTelemetry Authors
+// SPDX-License-Identifier: Apache-2.0
+
+package datapointstorage // import "github.com/open-telemetry/opentelemetry-collector-contrib/processor/metricstarttimeprocessor/internal/datapointstorage"
+
+import (
+	"sync"
+	"time"
+)
+
+// Notes on garbage collection (gc):
+//
+// Resource level gc:
+// The collector will likely execute in a long running service whose lifetime may exceed
+// the lifetimes of many of the jobs that it is collecting from. In order to keep the StartTimeCache from
+// leaking memory for entries of no-longer existing jobs, the StartTimeCache needs to remove entries that
+// haven't been accessed for a long period of time.
+//
+// Timeseries-level gc:
+// Some resources that the collector is collecting from may export timeseries based on metrics
+// from other resources (e.g. cAdvisor). In order to keep the timeseriesMap from leaking memory for entries
+// of no-longer existing resources, the timeseriesMap for each resource needs to remove entries that haven't
+// been accessed for a long period of time.
+//
+// The gc strategy uses a standard mark-and-sweep approach - each time a timeseriesMap is accessed,
+// it is marked. Similarly, each time a timeseriesInfo is accessed, it is also marked.
+//
+// At the end of each StartTimeCache.Get(), if the last time the StartTimeCache was gc'd exceeds the 'gcInterval',
+// the StartTimeCache is locked and any timeseriesMaps that are unmarked are removed from the StartTimeCache
+// otherwise the timeseriesMap is gc'd
+//
+// The gc for the timeseriesMap is straightforward - the map is locked and, for each timeseriesInfo
+// in the map, if it has not been marked, it is removed otherwise it is unmarked.
+//
+// Alternative Strategies
+// 1. If the resource-level gc doesn't run often enough, or runs too often, a separate go routine can
+//    be spawned at StartTimeCache creation time that gc's at periodic intervals. This approach potentially
+//    adds more contention and latency to each scrape so the current approach is used. Note that
+//    the go routine will need to be cancelled upon Shutdown().
+// 2. If the gc of each timeseriesMap during the gc of the StartTimeCache causes too much contention,
+//    the gc of timeseriesMaps can be moved to the end of MetricsAdjuster().AdjustMetricSlice(). This
+//    approach requires adding 'lastGC' Time and (potentially) a gcInterval duration to
+//    timeseriesMap so the current approach is used instead.
+
+// Cache maps from a resource to a map of timeseries instances for the resource.
+type Cache struct {
```

Comment:
Is this struct going to be part of a public API? If not, would you please move it under internal?

Reply:
not sure what you mean, its already in the `internal` subdirectory

Reply:
oh ok. All good :)